spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraConfCache.scala

```diff
@@ -0,0 +1,204 @@
+            Option(clusterReadConf.get(c)).getOrElse(defaultConf)))
+      case _       => Option(tableReadConf.get(Seq(table, keyspace))).getOrElse(
+        Option(keyspaceReadConf.get(Seq(keyspace))).getOrElse(defaultConf))
+    }
+  }
+
+  private val clusterWriteConf = java.util.Collections.synchronizedMap(
+    new java.util.HashMap[String, WriteConf]())
+  private val keyspaceWriteConf = java.util.Collections.synchronizedMap(
+    new java.util.HashMap[Seq[String], WriteConf]())
+  private val tableWriteConf = java.util.Collections.synchronizedMap(
+    new java.util.HashMap[Seq[String], WriteConf]())
+
+  /** Add table level write configuration settings. Set cluster to None for a single cluster */
+  def addTableWriteConf(keyspace: String,
+                        table: String,
+                        conf: WriteConf,
+                        cluster: Option[String]) : Unit = {
+    cluster match {
+      case Some(c) => validateClusterName(c)
+        tableWriteConf.put(Seq(table, keyspace, c), conf)
+      case _       => tableWriteConf.put(Seq(table, keyspace), conf)
+    }
+  }
+
+  /** Add keyspace level write configuration settings. Set cluster to None for a single cluster */
+  def addKeyspaceLevelWriteConf(keyspace: String,
+                                writeConf: WriteConf,
+                                cluster: Option[String]) : Unit = {
+    cluster match {
+      case Some(c) => validateClusterName(c)
+        keyspaceWriteConf.put(Seq(keyspace, c),  writeConf)
+      case _       => keyspaceWriteConf.put(Seq(keyspace), writeConf)
+    }
+  }
+
+  /** Add cluster level write configuration settings */
+  def addClusterLevelWriteConf(cluster: String, conf: WriteConf) : Unit = {
+    validateClusterName(cluster)
+    clusterWriteConf.put(cluster, conf)
+  }
+
+  /** Remove table level write configuration settings */
+  def removeTableLevelWriteConf(keyspace: String, table: String, cluster: Option[String]) : Unit = {
+    cluster match {
+      case Some(c) => validateClusterName(c)
+        tableWriteConf.remove(Seq(table, keyspace, c))
+      case _       => tableWriteConf.remove(Seq(table, keyspace))
+    }
+  }
+
+  /** Remove cluster level write configuration settings */
+  def removeKeyspaceLevelWriteConf(keyspace: String, cluster: Option[String]) : Unit = {
+    cluster match {
+      case Some(c) => validateClusterName(c)
+        keyspaceWriteConf.remove(Seq(keyspace, c))
+      case _       => keyspaceWriteConf.remove(Seq(keyspace))
+    }
+  }
+
+  /** Remove cluster level write configuration settings */
+  def removeClusterLevelWriteConf(cluster: String) : Unit = {
+    clusterWriteConf.remove(cluster)
+  }
+
+  /** Get write configuration settings by the order of table level, keyspace level, cluster level, default settings */
+  def getWriteConf(keyspace: String,
+                   table: String,
+                   cluster: Option[String],
+                   defaultConf: WriteConf): WriteConf = {
```

Comment:
Method parameters indentation wrong here. When splitting parameters into separate lines, you should use normal indent of 4 spaces. You should not align them to the method name, because when the method name get changed by performing "rename" refactoring, the alignment gets broken and you might not notice:
```scala
def getWriteConf(
    keyspace: String,
    table: String,
    cluster: Option[String],
    defaultConf: WriteConf): WriteConf = {
```